### PR TITLE
[needs help but nearly there] Component groups and data assignments

### PR DESCRIPTION
**Pull request by philastrophist**
1. Adds the ability to group components into groups whose internal amplitude ratios must remain the same (but the total weight of the group may change with respect to others). This allows us to fix one population's density and then fit another dataset that contains that population plus others. 
2. Adds the ability to assign data points to components when their labels are known

What little C I know has now been learnt from your codebase... So I think I've been doing well up to now, but I've reached an impasse where I now need help. 

I've added this snippet at the end of `proj_EM_step.c` and it compiles but immediately falls over when I run my test script.
```
  // normalise groups - within each group:
    if (gsl_matrix_get(groups, 0, 0) > -1) {
        double group_weight;
        for (int gg = 0; gg != ngroups; ++gg) {
            double group_sum = 0;
            double alpha_sum = 0;
            for (int kk = 0; kk != K; ++kk) {
                group_weight = gsl_matrix_get(groups, gg, kk);
                group_sum += group_weight;
                if (group_weight > 0) {
                    alpha_sum += gaussians->alpha;;
                }
                ++gaussians;
            }
            gaussians -= K;
            for (int kk = 0; kk != K; ++kk) {
                group_weight = gsl_matrix_get(groups, gg, kk);
                if (*(fixamp++) == false) {
                    if (group_weight > 0) {
                        gaussians->alpha = group_weight / group_sum * alpha_sum;
                    }
                }
                ++gaussians;
            }
            gaussians -= K;
            fixamp -= K;
        }
    }
```
It finishes with seg faults or free pointer errors with no obvious cause (at least to me). 
And on top of that, it doesn't do what it's supposed to anyway (the ratios of the component weights don't remain set).

I've attached the script I'm using to test.
I'd be really grateful if you could take a look, thanks @jobovy !

```
import numpy as np
from extreme_deconvolution import extreme_deconvolution \
    as bovyXD
from sklearn.mixture import GaussianMixture as skl_GMM

n_components = 4
X = np.concatenate([np.random.multivariate_normal([0, 0], np.eye(2), size=200),
                    np.random.multivariate_normal([10, 10], np.eye(2), size=200),
                    np.random.multivariate_normal([20, 20], np.eye(2), size=200)])
# deliberately made a bad model so it has to do some work 
Xerr = np.zeros(X.shape + (X.shape[1],))
Xerr[:] = np.eye(2) * 0.01


tmp_gmm = skl_GMM(n_components, max_iter=10,
                  covariance_type='full', random_state=0)
tmp_gmm.fit(X)
mu = tmp_gmm.means_
weights = tmp_gmm.weights_
V = tmp_gmm.covariances_
groups = np.array([[True, True, False, False],
                  [False, False, True, True]])  # split into two groups

print(weights[1]/weights[0], weights[3]/weights[2])

logl = bovyXD(X, Xerr, weights, mu, V,
              tol=1e-5, maxiter=10000, w=1e-6, groups=groups)

print(weights[1]/weights[0], weights[3]/weights[2])  # this should be the same as pre-fitting
print(mu)
print(weights)
print(V)
```

### CODE REVIEW

**Comment by jobovy** [comment]
Hi, thanks for your contribution! Looking through the code, I don't see any obvious reason for a segfault, but it's probably difficult to spot because it's so much code. Generally, a segfault means that you are manipulating and evaluating a pointer outside of its allocated range. My own strategy for finding the error is to just add a lot of print statements to check that everything is as I expect it, but there are probably better ways to do this as well 😄 .

I'm not entirely sure what you are trying to do and how it meshes with the algorithm as implemented. To ever merge this, I would have to see a write-up of the math and a proof that the algorithm still maximizes the likelihood.
